File: scripts/audit-recipe-images.mjs
import { createHash } from 'node:crypto'
import { readFileSync } from 'node:fs'
import { join } from 'node:path'

const root = process.cwd()
const recipes = readFileSync(join(root, 'src/data/recipes.ts'), 'utf8')
const catalog = readFileSync(join(root, 'src/data/catalogRecipes.ts'), 'utf8')
const featured = [...recipes.matchAll(/image:\s*'\.\/images\/([^']+\.webp)'/g)].map((match) => match[1])
const catalogImages = [...catalog.matchAll(/cuisineId:\s*'([^']+)'/g)].map((match) => `catalog-${match[1]}.webp`)
const names = [...featured, ...catalogImages]

if (names.length !== 72) throw new Error(`expected 72 recipe images, found ${names.length}`)
if (new Set(names).size !== names.length) throw new Error('two recipes share one image path')

const hashes = new Map()
for (const name of names) {
  const data = readFileSync(join(root, 'public/images', name))
  if (data.length < 25_000) throw new Error(`image is unexpectedly small: ${name}`)
  const digest = createHash('sha256').update(data).digest('hex')
  if (hashes.has(digest)) throw new Error(`exact duplicate images: ${hashes.get(digest)} / ${name}`)
  hashes.set(digest, name)
}

console.log(`audited ${names.length} unique recipe image paths and SHA-256 hashes`)

File: scripts/audit-recipe-images.py
"""Fail CI on missing, exact-duplicate, or near-duplicate recipe imagery."""

from __future__ import annotations

import hashlib
import itertools
import re
from pathlib import Path

from PIL import Image


ROOT = Path(__file__).resolve().parents[1]
IMAGES = ROOT / "public" / "images"
RECIPES = (ROOT / "src" / "data" / "recipes.ts").read_text(encoding="utf-8")
CATALOG = (ROOT / "src" / "data" / "catalogRecipes.ts").read_text(encoding="utf-8")

featured = re.findall(r"image:\s*'\.\/images\/([^']+\.webp)'", RECIPES)
catalog = [f"catalog-{item}.webp" for item in re.findall(r"cuisineId:\s*'([^']+)'", CATALOG)]
files = [IMAGES / name for name in featured + catalog]

if len(files) != 72:
    raise SystemExit(f"expected 72 recipe images, found {len(files)}")
if len({path.name for path in files}) != len(files):
    raise SystemExit("two recipes resolve to the same image path")


def dhash(path: Path, size: int = 16) -> int:
    with Image.open(path) as image:
        gray = image.convert("L").resize((size + 1, size), Image.Resampling.LANCZOS)
        pixels = list(gray.getdata())
        bits = [pixels[row * (size + 1) + col + 1] > pixels[row * (size + 1) + col] for row in range(size) for col in range(size)]
    value = 0
    for bit in bits:
        value = (value << 1) | bit
    return value


records: list[tuple[Path, str, int]] = []
for path in files:
    if not path.exists():
        raise SystemExit(f"missing image: {path.name}")
    with Image.open(path) as image:
        width, height = image.size
        if min(width, height) < 500 or not 0.85 <= width / height <= 1.85:
            raise SystemExit(f"unusable dimensions for {path.name}: {image.size}")
    digest = hashlib.sha256(path.read_bytes()).hexdigest()
    records.append((path, digest, dhash(path)))

for left, right in itertools.combinations(records, 2):
    if left[1] == right[1]:
        raise SystemExit(f"exact duplicate: {left[0].name} / {right[0].name}")
    distance = (left[2] ^ right[2]).bit_count()
    if distance <= 5:
        raise SystemExit(f"near duplicate (dHash {distance}): {left[0].name} / {right[0].name}")

print(f"audited {len(records)} unique recipe images; dimensions and hashes passed")
